### ISSUE TITLE
Mismatch between XLA Tensor and PyTorch Native Tensor Results for `torch.matmul` in FP16 Precision on NVIDIA GPU

### ISSUE THREAD

**Issue by lausannel** [pytorch/xla#7077]
## 🐛 Bug

I conducted a comparative analysis to test the accuracy of the `torch.matmul` operator with both XLA Tensors and PyTorch native Tensors at **FP16 precision**. For tensors with shapes [1, x] and [x, y], there appears to be a discrepancy of the order of **1e-4** in the computed results between XLA and PyTorch.

Specifically, when using FP16 precision, the mismatch between XLA Tensor and PyTorch Tensor results for the `torch.matmul` operation becomes evident, while in BF16 precision, the results between the two match exactly.

## To Reproduce

You can use the following code `op_example.py` to reproduce the results

```python
import os

import torch
import torch.nn as nn
import torch.optim as optim
import torch_xla
import torch_xla.core.xla_model as xm
import pandas as pd
import argparse
import random
import numpy as np
os.environ['PJRT_DEVICE'] = os.getenv('PJRT_DEVICE', 'CUDA')

seed = 42
torch.manual_seed(seed)
torch.cuda.manual_seed_all(seed)
torch.manual_seed(seed)
random.seed(seed)
np.random.seed(seed)
torch.use_deterministic_algorithms(True)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
os.environ['CUBLAS_WORKSPACE_CONFIG']=':4096:8'
os.environ['TF_CUDNN_DETERMINISTIC']='1'
os.environ['TF_DETERMINISTIC_OPS']='1'
xm.set_rng_state(seed)
torch_xla._XLAC._xla_set_use_full_mat_mul_precision(
    use_full_mat_mul_precision=True)
torch.set_printoptions(precision=10)


def parse_args():
    parser = argparse.ArgumentParser(description="Train Arguments")
    parser.add_argument("--input_size", type=int, default=5)
    parser.add_argument("--batch_size", type=int, default=1)
    parser.add_argument("--output_size", type=int, default=4)
    parser.add_argument("--fp16", action="store_true")
    parser.add_argument("--bf16", action="store_true")

    return parser.parse_args()

args = parse_args()

use_xla = os.environ.get("USE_ACC", 0)
device = xm.xla_device() if use_xla else torch.device('cuda')

input_size = args.input_size
output_size = args.output_size
batch_size = args.batch_size

params = torch.rand(input_size, output_size)
input = torch.randn(batch_size, input_size)

input = input.to(device)
with torch.amp.autocast(device_type='cuda', dtype=torch.float16 if args.fp16 else torch.bfloat16):
    output = torch.matmul(input, params.to(device))

if use_xla:
    print(f"XLA: {output}")
else:
    print(f"CUDA: {output}")
```
Use the following command to produce the result
1. XLA FP16
    ```shell
    USE_ACC=1 python op_example.py --fp16
     ```
2. CUDA FP16
    ```shell
    python op_example.py --fp16
    ```
3. XLA BF16
    ```
    USE_ACC=1 python op_example.py --bf16
    ```
4. CUDA BF16
    ```
    python op_example.py --bf16
    ```

## Expected behavior

The results for the commands are as follows:
```shell
# XLA FP16
XLA: tensor([[-0.5117187500, -0.4033203125, -0.0650634766,  0.4687500000]],
       device='xla:0', dtype=torch.float16)
# CUDA FP16
CUDA: tensor([[-0.5122070312, -0.4028320312, -0.0657348633,  0.4687500000]],
       device='cuda:0', dtype=torch.float16)
# XLA BF16
XLA: tensor([[-0.5117187500, -0.4003906250, -0.0673828125,  0.4667968750]],
       device='xla:0', dtype=torch.bfloat16)
# CUDA BF16
CUDA: tensor([[-0.5117187500, -0.4003906250, -0.0673828125,  0.4667968750]],
       device='cuda:0', dtype=torch.bfloat16)
```
The results demonstrate perfect agreement when using BF16 precision; however, with FP16 precision, we observe a mismatch of approximately 1e-4.

## Environment

 - Reproducible on XLA backend [CPU/TPU/CUDA]: CUDA
 - torch_xla version: 2.3.0

## Additional context

The current investigation has yielded the following findings:

1. When the first dimension (batch_size) of the initial tensor is set to 1, XLA does not compile the MatMul operation into GEMM; instead, it compiles it into an Elementwise kernel (Fused Reduce). Conversely, when the batch_size is greater than 1, torch.matmul is compiled into a cuBLAS GEMM call.
2. There are notable differences in the casting sequence between the HLO graphs generated for BF16 and FP16 precision. Observations from the BF16 HLO graph indicate that a cast to FP32 precedes the multiplication operation.
    ```C
    broadcast.3.5 = f32[5,4]{1,0} broadcast(convert.12.7), dimensions={0}
    param_0.22 = f32[5,4]{1,0} parameter(0)
    convert.6.3 = bf16[5,4]{1,0} convert(param_0.22)
    convert.10.3 = f32[5,4]{1,0} convert(convert.6.3)
    multiply.2.3 = f32[5,4]{1,0} multiply(broadcast.3.5, convert.10.3)
    ```
    On the other hand, the following code of the FP16 HLO graph demonstrate that multiplication is performed first, followed by a cast to FP32.
    ```C
    multiply.2.5 = f16[5,4]{1,0} multiply(broadcast.2.7, convert.6.3)
    convert.7.3 = f32[5,4]{1,0} convert(multiply.2.5)
    ```
     It is hypothesized that the discrepancies in precision and resulting accuracy may stem from this variance in operation order.
---
<details>
<summary>HLO Graph for FP16</summary>

```C
HloModule SyncTensorsGraph.7, is_scheduled=true, entry_computation_layout={(f32[5,4]{1,0}, f32[1,5]{1,0})->(f16[1,4]{1,0})}, frontend_attributes={fingerprint_before_lhs="c29de62e49d87b59c55311b356f4d188"}

scalar_add_computation {
  scalar_rhs.1 = f32[] parameter(1)
  scalar_lhs.1 = f32[] parameter(0)
  ROOT add.1.0 = f32[] add(scalar_lhs.1, scalar_rhs.1)
}

fused_reduce {
  param_1.14 = f32[1,5]{1,0} parameter(1)
  convert.5.3 = f16[1,5]{1,0} convert(param_1.14)
  bitcast.10.7 = f16[5]{0} bitcast(convert.5.3)
  broadcast.2.7 = f16[5,4]{1,0} broadcast(bitcast.10.7), dimensions={0}
  param_0.15 = f32[5,4]{1,0} parameter(0)
  convert.6.3 = f16[5,4]{1,0} convert(param_0.15)
  multiply.2.5 = f16[5,4]{1,0} multiply(broadcast.2.7, convert.6.3)
  convert.7.3 = f32[5,4]{1,0} convert(multiply.2.5)
  constant_5 = f32[] constant(0)
  reduce.3 = f32[4]{0} reduce(convert.7.3, constant_5), dimensions={0}, to_apply=scalar_add_computation
  ROOT convert.8.1 = f16[4]{0} convert(reduce.3)
} // fused_reduce

ENTRY SyncTensorsGraph.7 {
  p1.3.0 = f32[1,5]{1,0} parameter(1)
  p0.1.0 = f32[5,4]{1,0} parameter(0)
  loop_reduce_fusion = f16[4]{0} fusion(p0.1.0, p1.3.0), kind=kLoop, calls=fused_reduce
  bitcast.23.0 = f16[1,4]{1,0} bitcast(loop_reduce_fusion)
  ROOT tuple.6.0 = (f16[1,4]{1,0}) tuple(bitcast.23.0)
}
```
</details>

<details>
<summary>HLO Graph for BF16</summary>

```C
HloModule SyncTensorsGraph.7, is_scheduled=true, entry_computation_layout={(f32[5,4]{1,0}, f32[1,5]{1,0})->(bf16[1,4]{1,0})}, frontend_attributes={fingerprint_before_lhs="5d48e3f28f2dc34587fab7e6c9da0105"}

scalar_add_computation {
  scalar_rhs.1 = f32[] parameter(1)
  scalar_lhs.1 = f32[] parameter(0)
  ROOT add.1.0 = f32[] add(scalar_lhs.1, scalar_rhs.1)
}

fused_reduce {
  param_1.11 = f32[1,5]{1,0} parameter(1)
  convert.5.4 = bf16[1,5]{1,0} convert(param_1.11)
  bitcast.10.10 = bf16[5]{0} bitcast(convert.5.4)
  convert.12.7 = f32[5]{0} convert(bitcast.10.10)
  broadcast.3.5 = f32[5,4]{1,0} broadcast(convert.12.7), dimensions={0}
  param_0.22 = f32[5,4]{1,0} parameter(0)
  convert.6.3 = bf16[5,4]{1,0} convert(param_0.22)
  convert.10.3 = f32[5,4]{1,0} convert(convert.6.3)
  multiply.2.3 = f32[5,4]{1,0} multiply(broadcast.3.5, convert.10.3)
  constant_6 = f32[] constant(0)
  reduce.3 = f32[4]{0} reduce(multiply.2.3, constant_6), dimensions={0}, to_apply=scalar_add_computation
  ROOT convert.8.1 = bf16[4]{0} convert(reduce.3)
} // fused_reduce

ENTRY SyncTensorsGraph.7 {
  p1.3.0 = f32[1,5]{1,0} parameter(1)
  p0.1.0 = f32[5,4]{1,0} parameter(0)
  loop_reduce_fusion = bf16[4]{0} fusion(p0.1.0, p1.3.0), kind=kLoop, calls=fused_reduce
  bitcast.23.0 = bf16[1,4]{1,0} bitcast(loop_reduce_fusion)
  ROOT tuple.6.0 = (bf16[1,4]{1,0}) tuple(bitcast.23.0)
}
```
</details>

**Comment by lausannel** [comment]
Hi there! Any updates on this issue? Your assistance would be greatly appreciated. Thank you!

**Comment by vanbasten23** [comment]
Let me check

**Comment by vanbasten23** [comment]
Is the question about amp or in general?
I commented out the amp part https://gist.github.com/vanbasten23/2310184d4f6bf6047f8d7b49dd72493d and got the same conclusion as yours: the results on fp16 don't match. So it seems it's not an AMP issue.

The HLO looks similar
fp16:
```
ENTRY %IrToHlo.7 (p0.1: f32[5,4], p1.3: f32[1,5]) -> (f16[1,4]) {
  %p1.3 = f32[1,5]{1,0} parameter(1)
  %convert.4 = f16[1,5]{1,0} convert(f32[1,5]{1,0} %p1.3)
  %p0.1 = f32[5,4]{1,0} parameter(0)
  %convert.2 = f16[5,4]{1,0} convert(f32[5,4]{1,0} %p0.1)
  %dot.5 = f16[1,4]{1,0} dot(f16[1,5]{1,0} %convert.4, f16[5,4]{1,0} %convert.2), lhs_contracting_dims={1}, rhs_contracting_dims={0}, operand_precision={highest,highest}
  ROOT %tuple.6 = (f16[1,4]{1,0}) tuple(f16[1,4]{1,0} %dot.5)
}
```

bf16:
```
ENTRY %IrToHlo.7 (p0.1: f32[5,4], p1.3: f32[1,5]) -> (bf16[1,4]) {
  %p1.3 = f32[1,5]{1,0} parameter(1)
  %convert.4 = bf16[1,5]{1,0} convert(f32[1,5]{1,0} %p1.3)
  %p0.1 = f32[5,4]{1,0} parameter(0)
  %convert.2 = bf16[5,4]{1,0} convert(f32[5,4]{1,0} %p0.1)
  %dot.5 = bf16[1,4]{1,0} dot(bf16[1,5]{1,0} %convert.4, bf16[5,4]{1,0} %convert.2), lhs_contracting_dims={1}, rhs_contracting_dims={0}, operand_precision={highest,highest}
  ROOT %tuple.6 = (bf16[1,4]{1,0}) tuple(bf16[1,4]{1,0} %dot.5)
}
```

**Comment by lausannel** [comment]
@vanbasten23 
Thank you for your input. To clarify, this is indeed a general question regarding the behavior and not specifically related to AMP.

The HLO graphs I provided are from optimized computations. I exported these using the XLA flag setting as follows: `export XLA_FLAGS='--xla_dump_hlo_as_text --xla_dump_to=./hlo_test`.

**Comment by vanbasten23** [comment]
Right. With the XLA FLAGS, I got the post optimization dump https://gist.github.com/vanbasten23/81d9737e5ccbab23a635178641c4c6e6 for fp16 and https://gist.github.com/vanbasten23/2ff54ad3267330e38becaaf2320eef5c for bf16.

In fp16, the matmul operates on 2 fp16 operands. In bf16, it operates on 2 fp32 operands.

**Comment by vanbasten23** [comment]
I'm checking with XLA:GPU team currently. Will update you once I hear back.

**Comment by vanbasten23** [comment]
Per XLA:GPU, the relative difference of 0.09% for FP16 is expected and the difference is probably due to the difference in the order they are accumulated.

**Comment by lausannel** [comment]
@vanbasten23 Thank you for your help with this issue!
